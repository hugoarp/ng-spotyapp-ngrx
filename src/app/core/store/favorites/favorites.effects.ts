import { Injectable } from '@angular/core';
import { createEffect, ofType } from '@ngrx/effects';
import { Actions } from '@ngrx/effects';
import { SpotifyService } from 'src/app/core/services/spotify.service';
import { EMPTY, take } from 'rxjs';
import { map, mergeMap, catchError, tap } from 'rxjs/operators';
import {
  addFavoriteTrack,
  addTrackToState,
  loadedFavoritesTracks,
  loadFavoritesTracks,
  removeFavoriteTrack,
  removeTrackFromState,
} from './favorites.actions';
import { HotToastService } from '@ngneat/hot-toast';

@Injectable()
export class FavoritesTracksInfoEffects {
  constructor(
    private spotify: SpotifyService,
    private actions$: Actions,
    private toast: HotToastService
  ) {}

  loadFavoritesTracks$ = createEffect(() =>
    this.actions$.pipe(
      ofType(loadFavoritesTracks.type),
      mergeMap(() =>
        this.spotify.getUserFavorites().pipe(
          map((trackInfo) => ({
            type: loadedFavoritesTracks.type,
            trackInfo,
          })),
          catchError(() => EMPTY)
        )
      ),
      take(1)
    )
  );

  saveFavoritesTracks$ = createEffect(() =>
    this.actions$.pipe(
      ofType(addFavoriteTrack),
      mergeMap(({ trackInfo }) =>
        this.spotify.saveFavoriteTrack(trackInfo.id).pipe(
          map(() => ({
            type: addTrackToState.type,
            trackInfo,
          })),
          tap(() => {
            if (!trackInfo.favorite) {
              this.toast.show('AÃ±adido a favoritos', {
                icon: 'ðŸ’š',
                position: 'bottom-right',
              });
            } else {
              this.toast.show('Eliminado de favoritos', {
                icon: 'ðŸ’”',
                position: 'bottom-right',
              });
            }
          }),
          catchError(() => EMPTY)
        )
      )
    )
  );

  removeFavoritesTracks$ = createEffect(() =>
    this.actions$.pipe(
      ofType(removeFavoriteTrack),
      mergeMap(({ trackInfo }) =>
        this.spotify.deleteFavoriteTrack(trackInfo.id).pipe(
          map(() => ({
            type: removeTrackFromState.type,
            trackInfo,
          })),
          tap(() =>
            this.toast.show('Eliminado de favoritos', {
              icon: 'ðŸ’”',
              position: 'bottom-right',
            })
          ),
          catchError(() => EMPTY)
        )
      )
    )
  );
}
