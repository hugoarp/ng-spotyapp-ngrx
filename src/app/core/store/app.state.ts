import { ActionReducerMap } from '@ngrx/store';
import { TrackInfoState } from '../interfaces/track';
import { UserInfoState } from '../interfaces/user';
import { favoritesTrackInfoReducer } from './favorites/favorites.reducers';
import { recommendedTrackInfoReducer } from './recommended/recommended.reducers';
import { userInfoReducer } from './user/user.reducers';

export interface AppState {
  userInfo: UserInfoState;
  recommendedTrackInfo: TrackInfoState;
  favoritesTrackInfo: TrackInfoState;
}

export const ROOT_REDUCERS: ActionReducerMap<AppState> = {
  userInfo: userInfoReducer,
  recommendedTrackInfo: recommendedTrackInfoReducer,
  favoritesTrackInfo: favoritesTrackInfoReducer,
};
