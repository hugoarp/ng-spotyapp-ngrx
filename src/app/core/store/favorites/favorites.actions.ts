import { createAction, props } from '@ngrx/store';
import { TrackInfo } from '../../interfaces/track';

export const loadFavoritesTracks = createAction(
  '[Favorites] Load favorites tracks'
);

export const loadedFavoritesTracks = createAction(
  '[Favorites] Load favorites tracks success',
  props<{ trackInfo: TrackInfo[] }>()
);

export const addFavoriteTrack = createAction(
  '[Favorites] Add favorite track',
  props<{ trackInfo: TrackInfo }>()
);

export const addTrackToState = createAction(
  '[Favorites] Add favorite track to state',
  props<{ trackInfo: TrackInfo }>()
);

export const removeFavoriteTrack = createAction(
  '[Favorites] Remove favorite track',
  props<{ trackInfo: TrackInfo }>()
);

export const removeTrackFromState = createAction(
  '[Favorites] Remove favorite track from state',
  props<{ trackInfo: TrackInfo }>()
);
