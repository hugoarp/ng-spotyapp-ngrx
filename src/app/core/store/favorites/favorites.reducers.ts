import { createReducer, on } from '@ngrx/store';
import { TrackInfoState } from '../../interfaces/track';
import {
  addTrackToState,
  loadedFavoritesTracks,
  loadFavoritesTracks,
  removeTrackFromState,
} from './favorites.actions';

export const initialFavoritesTrackInfoState: TrackInfoState = {
  loading: false,
  tracks: [],
};

export const favoritesTrackInfoReducer = createReducer(
  initialFavoritesTrackInfoState,
  on(loadFavoritesTracks, (state) => {
    return { ...state, loading: true };
  }),
  on(loadedFavoritesTracks, (state, { trackInfo }) => {
    return { ...state, loading: false, tracks: trackInfo };
  }),
  on(addTrackToState, (state, { trackInfo }) => {
    return { ...state, tracks: [...state.tracks, trackInfo] };
  }),
  on(removeTrackFromState, (state, { trackInfo }) => {
    return {
      ...state,
      tracks: state.tracks.filter((track) => track.id !== trackInfo.id),
    };
  })
);
