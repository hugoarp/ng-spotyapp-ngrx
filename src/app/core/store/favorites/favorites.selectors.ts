import { createSelector } from '@ngrx/store';
import { TrackInfoState } from '../../interfaces/track';
import { AppState } from '../app.state';

export const selectFavoritesTracksFeature = (state: AppState) =>
  state.favoritesTrackInfo;

export const selectFavoritesTracks = createSelector(
  selectFavoritesTracksFeature,
  (state: TrackInfoState) => state.tracks
);

export const selectFavoritesTracksLoading = createSelector(
  selectFavoritesTracksFeature,
  (state: TrackInfoState) => state.loading
);
