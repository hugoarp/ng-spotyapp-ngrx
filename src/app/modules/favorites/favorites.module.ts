import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

import { FavoritesRoutingModule } from './favorites-routing.module';
import { FavoritesComponent } from './page/favorites.component';
import { OrganismsModule } from '@shared/components/organisms/organisms.module';
import { TemplatesModule } from '@shared/components/templates/templates.module';
import { MoleculesModule } from '@shared/components/molecules/molecules.module';

@NgModule({
  declarations: [FavoritesComponent],
  imports: [
    CommonModule,
    FavoritesRoutingModule,
    OrganismsModule,
    MoleculesModule,
    TemplatesModule,
  ],
  exports: [FavoritesComponent],
})
export class FavoritesModule {}
