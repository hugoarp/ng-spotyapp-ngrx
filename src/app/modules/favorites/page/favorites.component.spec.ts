import { ComponentFixture, TestBed } from '@angular/core/testing';
import { provideMockStore } from '@ngrx/store/testing';
import { FavoritesComponent } from './favorites.component';

describe('FavoritesComponent', () => {
  let component: FavoritesComponent;
  let fixture: ComponentFixture<FavoritesComponent>;
  let compiled: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [FavoritesComponent],
      providers: [provideMockStore({})],
    }).compileComponents();

    fixture = TestBed.createComponent(FavoritesComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  test('debe hacer match con el snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });
});
