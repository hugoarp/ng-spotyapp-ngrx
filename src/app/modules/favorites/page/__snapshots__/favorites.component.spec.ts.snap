// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoritesComponent debe hacer match con el snapshot 1`] = `
<div
  id="root1"
>
  
</div>
`;
